### DISCUSSION TITLE
Button-Designs and Layout-Questions

### DISCUSSION THREAD

**Discussion by TopperDEL** [StereoKit/StereoKit#118] [Q&A] [answered]
I would need these features for the UI. Maybe you can elaborate the effort or tell me where I can adjust them to my needs?

- Can buttons be in a "disabled" state?
- Can I change the font-family, font-style, font-size, font-alignment anywhere?
- Can I change the Background-color of a button?
- I need a Label-Kind-Element that look like a not-pushable button. I know you have a label, but it does not have it's text centered like the button have. And using Text.Add() does not place the text in the UI-layout.
- Can Text.Add() be surrounded by a "background-plate" (for better readability)? In this case it would be great to have a margin/padding-property and a border (with thickness, bordercolor, corner radius).

I know this all leads to a huge field of UI, Layouting and such where you could put all your time and effort in it without ever being ready. But maybe you could give me more insight or some basic building blocks to use. I already tried to understand how a button gets rendered - but most of it seems to happen on the NativeAPI-side that I do not really understand.

I'm not quite sure if it would be possible to create a flexible control like the Button in UWP/XAML - it basically has a Border, a Content and a clickable-component. If I would have this in StereoKit, I could disable click and place text in it to have my "label-kind-control" from above. But I could also maybe place another Stepper in it to render some totally different content within it.

This all is just to get something started. I'm happy to help here if you could provide me some guidance or "native-helpers" to achieve the parts of it. Thank you. :)

**Comment by maluoi** [comment, marked as answer]
The only tools I have for UI styling are currently on display [over here](https://github.com/maluoi/StereoKit/blob/master/Examples/StereoKitTest/Demos/DemoUISettings.cs), which is unfortunately not a ton! A couple things you mention might be pretty easy to do, like some additional control over font, style, alignment, and colors. A real flexible layout and visual styling system is a Big Task though.

Once I finish the v0.3 update, I'll sit down and think about the UI a bunch more, so definitely keep surfacing feedback and feature requests here! I don't have much in the way of sample code with using the UI building functions, and I wouldn't be surprised if they're even a bit stiff for building custom controls. I haven't given them enough of a workout yet!

This does start to get into some questions that I've largely been putting off until later, which is generally: what is the purpose of StereoKit's UI? I don't think that UI is actually a good responsibility for StereoKit's core, there's too much in the air about what a MR UI should even look like, let alone what the API should be! I want StereoKit to be useful for everyone, including those that want to take UI in a completely different direction.

But I built StereoKit's UI anyhow, because I felt that having even a rudimentary system in place provides a _huge_ amount of benefit to most developers, as most are not going to want to build that from scratch. An Immediate Mode system doesn't have much in the way of internal state either, which jives with my design philosophy, and this implementation is entirely build on top of StereoKit's systems, so anyone should be able to replicate or replace it using only the public API. One downside to Immediate Mode UIs, is that they're often difficult to style or layout nicely!

So I'm uncertain if StereoKit's current UI should be the actual UI system, or if it should be just an easy-to-use placeholder UI thing! I think a final system would also have some WYSIWYG features, some XML layout tools, or whatever the kids are doing these days. An interesting thought related to that, and kinda what I've been leaning towards lately, is that StereoKit's current UI could possibly be a layer underneath some more modern/advanced system. Use the upper layer for managing layout and styling properly, and the existing system for the interaction and rendering.

I dunno if that answers much of your questions here, but I'll definitely be referring back to this post when I'm reviewing things!

**Reply by TopperDEL** [reply]
That already helps alot! I'm with you if you say that StereoKit is more for the basic stuff an not about a UI-framework. It's good to have the basics at hand, but it might not need to be the goal for StereoKit to make every UI-use-case possible. There's always the possibility to build everything on my own which is much more complicated in languages like XAML - so that's ok here.

I think I roll my own UI for that case. I do not need much currently, but it needs to look exactly the way I want it. I just have to digg a little deeper in some basic helpers you provide. Not sure if you can give me some more hints regarding these functionalities:
- a good way to make something touchable (to create my own button)
- a possibility to use another font (if possible)
- how I may create a dynamically scaling "box" containing text with wrapping, that has a border. I guess the  Text.Size-function may help here.

**Reply by maluoi** [reply]
So I do have an interesting example of a "painting palette menu" [over here](https://github.com/maluoi/StereoKit-PaintTutorial), check out PaletteMenu.cs for the code there. This uses manual layout and mesh 'buttons' to create a more interesting UI object.

For different Fonts, check out [TextStyle](https://stereokit.net/Pages/Reference/Text/MakeStyle.html), and there's a tiny bit more of that over in the [source file](https://github.com/maluoi/StereoKit/blob/master/Examples/StereoKitTest/Demos/DemoText.cs) for that page.

And yes, Text.Size is exactly what you need for dynamically sizing a box :)

**Reply by TopperDEL** [reply]
I managed to create my own Buttons and dynamic-scaling-textblocks. Thanks for the additional info!
